Blackjack/Game/BlackjackGame.m

```diff
@@ -0,0 +1,262 @@
+const NSInteger MINIMUM_ROUND_BET = 5;
+
+// MARK: - Init
+
+- (instancetype)initWithNumberOfPlayers:(NSInteger)numberOfPlayers delegate:(id<BlackjackGameDelegate>)delegate {
+  return [self initWithDeck:[PlayingCardDeck new] numberOfPlayers:numberOfPlayers delegate:delegate];
+}
+
+- (instancetype)initWithDeck:(Deck *)deck numberOfPlayers:(NSInteger)numberOfPlayers
+                    delegate:(id<BlackjackGameDelegate>)delegate {
+  self = [super init];
+  if (self) {
+    NSMutableArray *newPlayers = [NSMutableArray arrayWithCapacity:numberOfPlayers];
+    
+    for (int i=0; i<numberOfPlayers; i++) {
+      NSString *name = [NSString stringWithFormat:@"Player #%d", i+1];
+      Player *player = [[Player alloc] initWithName:name chips: MINIMUM_BUYIN delegate:self];
+      [newPlayers addObject: player];
+    }
+    _players = newPlayers;
+    _currentPlayer = newPlayers.firstObject;
+    _deck = deck;
+    _delegate = delegate;
+    _bets = [NSMutableDictionary new];
+    _dealer = [[Player alloc] initWithName:@"Dealer" chips: 0 delegate:self];
+    _state = IN_GAME;
+  }
+  return self;
+}
+
+- (void)setState:(GameState)state {
+  _state = state;
+  [self.delegate handleChangesforNewState: state];
+}
+
+// MARK: - Player Filters
+
+- (NSInteger)getBetAmountForPlayer:(Player *)player {
+  return [self.bets valueForKey:player.identifier].integerValue;
+}
+
+- (NSArray<Player *> *)currentActivePlayers {
+  NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(Player *player, NSDictionary *bindings) {
+    return player.state == PLAYING;
+  }];
+  return [self.players filteredArrayUsingPredicate:predicate];
+}
+
+- (NSArray<Player *> *)playersAvailableForNextFound {
+  NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(Player *player, NSDictionary *bindings) {
+    return player.chips >= MINIMUM_ROUND_BET;
+  }];
+  return [self.players filteredArrayUsingPredicate:predicate];
+}
+
+// MARK: - Game Logic
+
+- (void)collectDecision:(Decision)decision {
+  BOOL isLastPlayer = self.currentPlayer == [self currentActivePlayers].lastObject;
+  
+  switch (decision) {
+    case HIT: {
+      [self.currentPlayer acceptNewCard:[self.deck drawRandomCardWithFaceUp:YES]];
+      
+      if (self.currentPlayer.state == PLAYING) {
+        return;
+      }
+      break;
+    }
+      
+    case SURRENDER:
+      [self.currentPlayer setState:BUST];
+      
+    case STAND: {
+      break;
+    }
+  }
+  
+  if (isLastPlayer) {
+    [self finlizeRound];
+  } else {
+    [self nextPlayer];
+  }
+}
+
+- (void)collectBet:(NSInteger)amount {
+  if (self.currentPlayer.chips >= amount) {
+    [self.bets setObject:[NSNumber numberWithInteger:amount] forKey:self.currentPlayer.identifier];
+    [self.currentPlayer collectBet:amount];
+    [self nextPlayer];
+  } else {
+    NSAssert(self.currentPlayer.chips < amount, @"Player should never get to this point");
+  }
+  
+  if (self.bets[self.players.lastObject.identifier]) {
```

Comment:
Why are you checking this with regards to the bet?
In reality you are checking if you reached the last player, which should be done by checking the `self.currentPlayer == self.players.lastObject`

Reply:
Also, since this is not part of your bet collection, this should be done outside of this function

Reply:
I encapsulated some logic, but the if is still there, I need a way to know that bets are over.